hopsworks-common/src/main/java/io/hops/hopsworks/common/hosts/ServiceDiscoveryController.java

```diff
@@ -0,0 +1,110 @@
+ * If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package io.hops.hopsworks.common.hosts;
+
+import com.logicalclocks.servicediscoverclient.Builder;
+import com.logicalclocks.servicediscoverclient.ServiceDiscoveryClient;
+import com.logicalclocks.servicediscoverclient.exceptions.ServiceDiscoveryException;
+import com.logicalclocks.servicediscoverclient.exceptions.ServiceNotFoundException;
+import com.logicalclocks.servicediscoverclient.resolvers.Type;
+import com.logicalclocks.servicediscoverclient.service.Service;
+import com.logicalclocks.servicediscoverclient.service.ServiceQuery;
+import io.hops.hopsworks.common.util.Settings;
+
+import javax.ejb.EJB;
+import javax.ejb.Lock;
+import javax.ejb.LockType;
+import javax.ejb.Singleton;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Singleton
+@TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
+public class ServiceDiscoveryController {
+  
+  public enum SERVICE {
+    LIVY("livy"),
+    ZOOKEEPER_CLIENT("client.zookeeper"),
+    SPARK_HISTORY_SERVER("sparkhistoryserver"),
+    HTTP_RESOURCEMANAGER("http.resourcemanager"),
+    HIVE_SERVER_PLAIN("hiveserver2-plain.hive"),
+    HIVE_SERVER_TLS("hiveserver2-tls.hive"),
+    RPC_NAMENODE("rpc.namenode");
+    
+    private String name;
+    SERVICE(String name) {
+      this.name = name;
+    }
+    
+    public String getServiceName() {
+      return this.name;
+    }
+  }
+  
+  private static final String CONSUL_SERVICE_TEMPLATE = "%s.service.%s";
+  
+  @EJB
+  private Settings settings;
+  
+  @Lock(LockType.READ)
+  public String constructServiceFQDN(SERVICE service) {
+    String serviceName = service.getServiceName();
+    if (serviceName.endsWith(".")) {
+      serviceName = serviceName.substring(0, serviceName.length() - 1);
+    }
+    return String.format(CONSUL_SERVICE_TEMPLATE, serviceName, settings.getServiceDiscoveryDomain());
+  }
+  
+  @Lock(LockType.READ)
+  public Stream<Service> getService(Type resolverType, ServiceQuery serviceQuery)
+      throws ServiceDiscoveryException {
+    ServiceDiscoveryClient client = getClient(resolverType);
+    try {
+      return client.getService(serviceQuery);
+    } finally {
+      if (client != null) {
+        client.close();
+      }
+    }
+  }
+  
+  @Lock(LockType.READ)
+  public Service getAnyAddressOfServiceWithDNS(SERVICE serviceName) throws ServiceDiscoveryException {
+    ServiceQuery serviceQuery = ServiceQuery.of(constructServiceFQDN(serviceName), Collections.emptySet());
+    Optional<Service> serviceOpt = getService(Type.DNS, serviceQuery).findAny();
+    if (!serviceOpt.isPresent()) {
+      throw new ServiceNotFoundException("Could not find service with: " + serviceQuery);
+    }
+    return serviceOpt.get();
+  }
+  
+  private ServiceDiscoveryClient getClient(Type type) throws ServiceDiscoveryException {
+    switch (type) {
```

Comment:
switch we only one case is a bit redundant

Reply:
I added the rest of `Types` which throw a `NotImplementedException`